pkg/controller/generic/transform/controller.go

```diff
@@ -0,0 +1,217 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package transform
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/siderolabs/gen/xerrors"
+	"go.uber.org/zap"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+)
+
+// Controller provides a generic implementation of a controller which implements controller transforming Input resources into Output resources.
+type Controller[Input generic.ResourceWithRD, Output generic.ResourceWithRD] struct {
+	mapFunc       func(Input) Output
+	transformFunc func(context.Context, controller.Reader, *zap.Logger, Input, Output) error
+	generic.NamedController
+	options ControllerOptions
+}
+
+// NewController creates a new TransformController.
+func NewController[Input generic.ResourceWithRD, Output generic.ResourceWithRD](
+	name string,
+	mapFunc func(Input) Output,
+	transformFunc func(context.Context, controller.Reader, *zap.Logger, Input, Output) error,
+	opts ...ControllerOption,
+) *Controller[Input, Output] {
+	var options ControllerOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &Controller[Input, Output]{
+		NamedController: generic.NamedController{
+			ControllerName: name,
+		},
+		mapFunc:       mapFunc,
+		transformFunc: transformFunc,
+		options:       options,
+	}
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *Controller[Input, Output]) Inputs() []controller.Input {
+	var (
+		input  Input
+		output Output
+	)
+
+	inputKind := controller.InputWeak
+
+	if ctrl.options.inputFinalizers {
+		inputKind = controller.InputStrong
+	}
+
+	inputs := []controller.Input{
+		{
+			Namespace: input.ResourceDefinition().DefaultNamespace,
+			Type:      input.ResourceDefinition().Type,
+			Kind:      inputKind,
+		},
+		{
+			Namespace: output.ResourceDefinition().DefaultNamespace,
+			Type:      output.ResourceDefinition().Type,
+			Kind:      controller.InputDestroyReady,
+		},
+	}
+
+	return append(inputs, ctrl.options.extraInputs...)
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *Controller[Input, Output]) Outputs() []controller.Output {
+	var output Output
+
+	return []controller.Output{
+		{
+			Type: output.ResourceDefinition().Type,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocognit,gocyclo,cyclop
+func (ctrl *Controller[Input, Output]) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
```

Comment:
Run is incredibly hard to read. Maybe we should finally split into separate functions, since we are already refactoring it?

Reply:
yeah, I agree, next problem is to figure out _how_